### ISSUE TITLE
Frequency functionality for datasets

### ISSUE THREAD

**Issue by susanodd** [Signbank/Global-signbank#756]
At the moment, frequency is only available for the NGT dataset. Expand this functionality to other datasets.

This extends #354, which focused on display of frequency data, once the data is already available. That issue was opened in 2017, before datasets existed. It had a slightly different understanding of datasets, for example gender, regions, and ages.

**Comment by susanodd** [comment]
@ocrasborn, what shall the naming convention be for corpora vs datasets?
At the moment, there is exactly one Corpus object:
### id,       name,        description,        speakers_are_cross_referenced
`1,        'NGT',          'Corpus NGT',      True`

The EAF files are Document objects that link to this Corpus via a corpus_id field.

On the Dataset Manager page, the user can upload EAF files for a dataset.
Shall the Dataset acronym be used as the Corpus name?

**Comment by susanodd** [comment]
Here is a first go at a Dataset Corpus Frequency Overview.
This shows that of NGT.
I put the meta file and the EAF files (scroll panel) there that are uploaded on the Manage page.

Then the Corpus (it's just the name; all the other objects associated with it are linked to it).

Then the Speakers (in a scroll panel). These have been retrieved from the GlossFrequency objects that are stored for the Corpus, its Documents, and Speakers objects. [This is what is actually stored shown in this panel; it has not been retrieved from the metadata file.]

So the idea would be to display an overview and make handles available for other datasets that have EAF files and metadata.
Obviously, there needs to be a button to generate speakers and parse the EAF files (during the night).


![dataset-corpus-overview](https://user-images.githubusercontent.com/24272544/123259369-ad53ef80-d4f4-11eb-83a4-e868487dbf4c.png)

**Comment by ocrasborn** [comment]
@susanodd , yes, using the Dataset acronym for corpus name is fine.

Note that for Signbank's purposes, we don't really care whether the uploaded EAF files constitute a specific corpus with a specific identifier; it could be they combine EAF files from multiple datasets, or adult and child data, or whatever. It's up to them to spell that out in the description of their dataset in Signbank. We leave that up to the dataset managers. They upload EAF files, and Signbank calculates frequencies of glosses in those files.

**Comment by susanodd** [comment]
Okay, I'll add that functionality. The Document objects refer to a Corpus object.
I just checked, and that does not preclude that a particular EAF file may be uploaded to more than one dataset.

**Comment by susanodd** [comment]
Would you like separate functionality for documents?
This to avoid uploading them multiple times.

**Comment by susanodd** [comment]
@ocrasborn, here is what I'm not sure about.
Suppose you have some EAF files for a particular sign language or project.
To what extent / can these be used by multiple datasets that are for the same sign language?

As the upload of a metadata and eafs is currently implemented, they are uploaded for a dataset by a dataset manager. 
The metadata file gets the filename e.g., Acronym_metadata.csv and is stored in the folder metadata_eafs.
The EAF files are currently uploaded for a dataset and stored in the (sub)folder eafs/Acronym.

Is functionality needed so a user can "choose" documents (for a particular sign language) to associate with their corpus dataset?

And to what extent are Speaker codes created and used?

**Comment by ocrasborn** [comment]
Hi @susanodd , we want to keep this functionality and the associated files strictly linked to single datasets. If there would ever be multiple datasets for one and the same language, they would have their own set of EAF files and their own metadata file. So there is no need for cross-referencing or selecting files, the user posts the files in the upload function (and if there's new annotations they upload files with the same names that should overwrite the previous ones) and the system works with what is in the folder on the server at that particular time. 

In other words, the function can be blind for the property 'language', it needn't be in the EAF or metadata file names, we just need to keep the files separate on the server like we now do for the video files, for instance.

**Comment by susanodd** [comment]
Okay, thanks for the description!

**Comment by susanodd** [comment]
I've made it so that if there is no corpus, it gives information about creating one.
Otherwise it shows the (retrieved from the database) speakers and documents for the corpus, in scroll panels.
(Both to be made more elegant, with buttons for creating the corpus.)

![corpus-overv0ew-katakolok](https://user-images.githubusercontent.com/24272544/123412181-9a532500-d5b1-11eb-9fda-9cee24611f38.png)

![corpus-overview-ngt](https://user-images.githubusercontent.com/24272544/123412199-a0490600-d5b1-11eb-94be-02b15024cbe5.png)

**Comment by ocrasborn** [comment]
That looks great @susanodd ! Esp. the implicit help that the first screen delivers on what can be in the metadata. (This yet needs to be described properly in the manual, but it never hurts to remind people in the interface.)
Since most corpora will be large, can you make both scrolling panels a bit larger? Say, 10 lines instead of 4?
One question: what is the role of the 'Corpus' field, which is generated by the system by duplicating the dataset Acronym? I would suggest just hiding it from the user even if you need it in the code (although there too it will double with Acronym, no?).

**Comment by susanodd** [comment]
Okay, yes, the Corpus field is a bit silly looking. But it is handy for showing there isn't one.

I will make the scroll panels larger. I'm going to format them to look more like tables.

**Comment by susanodd** [comment]
Here are the improved scroll panels.

![corpus-ngt-overview](https://user-images.githubusercontent.com/24272544/125448190-195633d6-7162-4ff4-8d21-beb99dbb1cbb.png)

**Comment by ocrasborn** [comment]
Looks good @susanodd ! (This is my last response before a long summer break)

**Comment by susanodd** [comment]
@Woseseltops @vanlummelhuizen Do you guys recall when the Corpus data for NGT was first created?
(The Speaker, Document, GlossFrequency objects.)
There is functionality for doing this, currently only for NGT (CNGT).

I am adding this for other datasets. (This issue.)
But I can't tell whether the current (NGT specific) code is actually being invoked? Is there a cron job doing this?
If so, I will create a new url.

I need to account for not accidentally causing the current (CNGT) code not to work as I enhance (other) code to allow a new corpus to be created.

**Comment by susanodd** [comment]
I've got some functionality implemented now for other datasets. (i.e., not NGT or corpus CNGT)

I've been checking the code via the "tests". I'm using dataset tstMH as a testing. I've uploaded a metadata file with two speakers, and one eaf file as a test file. It uses those two speakers.

The problem in doing so, the "eaf file" as actually an eaf file for CNGT and the glosses that appear in it are in NGT.

The test database, I've modified it so there are zero GlossFrequency, Speaker, Corpus, Document objects in the initial test database.
So how the test procedes, it creates a new corpus for tstMH using the metadata file and the eaf file.
However, I had to construct glosses in dataset tstMH by hand to match the gloss ids in the eaf file.
(This is only in the test database for dataset tstMH. Because this is actually totally not possible because the gloss ids are unique and those in the eaf file already belong to NGT.)

So as described, I can test the new functionality, but only for inside of tests. I have no way to create eaf files and speakers for tstMH. So I can't get any actual test data.

Does this sound okay?
It's not entirely possible to test out the new functionality for any real data.

Ideally, the development would need to be assisted on the live server. Perhaps adding a "remove and construct the corpus again" in order to account for unexpected things happening. Although as far as I can tell, it seems to be working.
The only worrying aspect is accidentally reusing Speaker identifiers in totally different corpus documents for other languages.

**Comment by ocrasborn** [comment]
@susanodd , I'm now on vacation (....) but can work with you creating some more realistic test data in the week of August 2

**Comment by susanodd** [comment]
So far I've been modifying a test EAF file. I modified some gloss ids in this file to match glosses in dataset tstMH.
(The test file is nonsense of course. But it parses and I can use it to write the implementation code.)

What I've come to realise is that the "Number of occurrences" data in the Gloss Detail View is hard coded to CNGT.
The regions are also hard coded.

In my fake Metadata CSV file, it also has regions. But for other sign languages the hard coded regions are not okay.

I've adjusted the Corpus page so you can now create a new corpus once the metadata file and at least one EAF file has been uploaded.
I'm working on this so you can also refresh the corpus if you've added new files or updated the EAF files.

This is where I came to realise that the Gloss fields tokNo and tokNoSgnr
need to be updated (filled in) for this other dataset using its EAF files.

This works as expected. But the "region hard coded" Gloss model fields need to be removed or made parametric, or perhaps a Region model.
The fields are visible on the Gloss Detail View in the Frequency Panel.
There is also the Frequency View where information about Speakers is displayed. (But not the number of occurrences.)

**Comment by ocrasborn** [comment]
Hi @susanodd , the hard-coded fields you see in the Detail View can be hidden/removed by now, since we now have the Frequency View showing the graphic display. Not sure about the fields in the database though, whether the data in the Frequency View tab still come from those. I gather they don't, that something was already changed in terms of the data model when the Frequency View was created -- you might want to discuss this with @Woseseltops before moving on. 
Still, let me know if you'd like me to create other test data for you.

**Comment by susanodd** [comment]
Thanks @ocrasborn. In the Frequency view, there is only shown attributes of the speakers. Not the number of occurrences of the gloss. It's true, the Frequency View uses the new model classes for frequency.
But the Detail view uses the old fields in Gloss. That's how I noticed there was a problem. I started actually making a real corpus with my fake test data. I'm working on a "Update Corpus" button now. At present there isn't a way to delete "bad" eaf files. Only overwrite them or upload new ones. So on my local machine I've been renaming my test eaf file to see what happens when I "update" (with a new routine). I tweaked the "update of tokNo in the gloss. That actually also needs attention if files can be deleted. For the new model, this isn't a problem, deleting a document, also deletes the frequency data,.
I'll see about moving the "tokNo" display onto the Frequency View where it should be instead, and using the new models.

**Comment by susanodd** [comment]
@ocrasborn on the Gloss Frequency View, the first panel is Gloss Region.
This data comes from the Gloss fields specific to NGT.
However, the design of the settings sets a setting:
https://github.com/Signbank/Global-signbank/blob/94dbe93ee7552b7d9cd19416ee29b86d5d18dd61/signbank/settings/server_specific/global_sb_new_aj.py#L62
You can see those are the things we have so far.
At the moment, these are "displayed per region" as defined by the hard-coded NGT fields of Gloss.

So the question is, do Regions need to stay there for other datasets? (These are only via the Speaker meta data.)
How do you wish to present the Occurences and Signers data for other datasets?
It should replace the section "Gloss Regions". Or dynamically generate regions from the Speaker information.

**Comment by susanodd** [comment]
Okay, the code has been adapted so the Gloss Regions section of Frequency View computes these based on the GlossFrequency model objects. The regions are retrieved via the referenced signers, for the gloss being viewed.

**Comment by susanodd** [comment]
Here's the frequency view again. This is for Gloss PAKKEN-A of NGT. (1206). I added some gloss information at the start because it's otherwise not clear what you're looking at.


![gloss-frequency-view-1206](https://user-images.githubusercontent.com/24272544/130223845-b1039331-46cb-410c-ad82-0add0f511edc.png)


In revising the code to get the relevant data out of the GlossFrequency objects to display I discovered the method in question was also being used in a so-called Lemma Frequency View. That view actually isn't attached to anything else. Apparently @Woseseltops and I were working on this before the lockdown started last year. I seem to recall it was his idea to see if this was something useful. There still aren't that many glosses in lemma groups. The Lemma Frequency View is accessed not via the Lemma ID, but via the Gloss IDs that have this lemma. Here is the Lemma Frequency View for the above gloss 1206.
@ocrasborn any idea where such a view should be attached to other pages? (This is on my computer. The page on Global has a problem with the swatches all being the same grey.) The code already exists, it's just not attached to the rest.

![lemma-frequency-view-1206](https://user-images.githubusercontent.com/24272544/130224324-f9db03a3-f0bd-4908-96dc-aa5c8b56895b.png)

**Comment by susanodd** [comment]
Here are some screenshots of the /datasets/frequency/<dataset_id> page (on my local server)
It can only be accessed by logged in users. The following screenshots are for a logged in user.

If there are no corpus documents, then there are instructions for creating one.

![datasets_frequency_no_files](https://user-images.githubusercontent.com/24272544/135079154-a4feeae1-8b5a-4eeb-8403-3559216904b1.png)

If the user is **not** in group Manager for this dataset, then the files are shown, but there is no Update button.

![datasets_frequency_no_update](https://user-images.githubusercontent.com/24272544/135079318-cc904715-f099-4d36-963e-16b3062e571a.png)

If the user is in group Manager for this dataset, then there is an Update (or Create button).

![datasets_frequency_yes_update](https://user-images.githubusercontent.com/24272544/135079750-a5593372-3909-4ad8-a5a9-338da20ddc75.png)

**Comment by susanodd** [comment]
@ocrasborn the functionality is all implemented. Plus, the Frequency View uses data from the Dataset's Corpus. (The corpus must exist to show the frequency data.)

One thing that needs to be ironed out is the testing.
I added testing for a bogus tstMH corpus, with a hacked EAF file and fake glosses.
I did this on my local computer to test corpus creation, so that all the stuff generated is temporary.
I then actually did it on the real database (on my server) to make sure it works.
If you "create a corpus again" is being ironed out.
It's possible to "import a meta file - speaker information" multiple times. It just updates them.
It's possible to "process EAF files" (via the Update button). and this reads them in an processes them again.
(Q: What to do about this? I have exactly two fake EAF files, so I'm not processing huge numbers of files.)
Also, it's a bit obscure when a file has been updated. The "date" is actually encoded inside the EAF file. It's not the same as the file system date. Is this the intended time stamp?

**Comment by susanodd** [comment]
> @susanodd , yes, using the Dataset acronym for corpus name is fine.
> 
> Note that for Signbank's purposes, we don't really care whether the uploaded EAF files constitute a specific corpus with a specific identifier; it could be they combine EAF files from multiple datasets, or adult and child data, or whatever. It's up to them to spell that out in the description of their dataset in Signbank. We leave that up to the dataset managers. They upload EAF files, and Signbank calculates frequencies of glosses in those files.

With the EAF files, the Gloss IDs are inside of the EAF file. That entails that the glosses are "found" in the dataset being processed.
This caused problems for testing, because the "fake EAF file" was for CNGT, and it contained Gloss IDs for NGT. But I wanted to use this EAF file for tstMH, and it generated errors because those Gloss IDs inside of the EAF file belong to NGT, not tstMH.

So I modified the "imposter" EAF and created a speaker meta data file for tstMH, and modified the "imposter" EAF file to refer to those new speakers (participants), and also modified the Gloss IDs inside the "imposter" EAF to refer to Glosses that exist inside of tstMH.

So I don't entirely see how a user can use EAF files from CNGT to upload them to a different Dataset.
What to do with 1) the speaker identifiers, 2) the gloss ids that exist inside the EAF file?

**Comment by susanodd** [comment]
@vanlummelhuizen are you available for brainstorming?
As we have it, a gloss only belongs to one dataset.
Does this need to be modified in order to support sharing of EAF files between different corpora?

**Comment by Woseseltops** [comment]
For completeness, we're having the architecture discussion about the relation between EAF files, corpora, datasets and speakers in #354.

> But I can't tell whether the current (NGT specific) code is actually being invoked? Is there a cron job doing this?

I think so yes, but unfortunately all cron jobs are gone, see #769 .

> I've been checking the code via the "tests". 

https://en.wikipedia.org/wiki/Test-driven_development :)

> Any opinions about how robust or valid such a test EAF needs to be? Should there be a full-fledged test dataset for this purpose?

The situation you describe seems like a good start... I would say this functionality is cool but not crucial, so test coverage does not have to be 100%

**Comment by susanodd** [comment]
Yes, there seems to be a cron job.
I found the function to update the counts in tools.py.
https://github.com/Signbank/Global-signbank/blob/94dbe93ee7552b7d9cd19416ee29b86d5d18dd61/signbank/tools.py#L1972

**Comment by susanodd** [comment]
I revised the Speaker identifiers to have `_`<_dataset.acronym_>
This is more or less straightforward.
I modified the tests to use testing to check the code during revision.
If a new metadata file is read, or a corpus is updated, then the Speaker identifiers are so revised.

So far these are the tests:

- the count of the speakers imported from the metadata file is the same as the length of the metadata file
- the count of the speakers after importing the metadata file a second time remains the same
- all the documents configured for the new corpus have existing eaf files
- all the glosses in the GlossFrequency objects exist.
- all the speakers in the GlossFrequency objects exist
- updating a corpus (with exactly the same eaf files) does not change the GlossFrequency objects (i.e., glosses referenced, speakers references, frequencies tabulated)

**Comment by susanodd** [comment]
Perhaps an additional button on the DatasetFrequency / Corpus Overview page that allows to Update Speakers (so merely reprocesses the metadata rather than also the eaf files.)
That way it would be easier to allow additional information about speakers or modify region specificity.

**Comment by susanodd** [comment]
This is what the interface looks like now for new corpora. Shown are a few different datasets on my local server.
I am using the real (local) database now rather than the test database.

Here is NGT. I processed all 2281 files on my server. It took quite a long time.
So to remedy this (when not done as nighttime scripts) I have added refresh buttons for each document row.
Also, a panel to show which documents the speaker appears in. These are intended to assist the user.

![corpus_overview_ngt](https://user-images.githubusercontent.com/24272544/140089454-f42c01e5-c228-428a-a222-f5d7d9821f76.png)

**Comment by susanodd** [comment]
Here is tstMH. This has only a few documents. I used the same speakers metafile as that of NGT, so it actually doesn't find very many glosses because it's NGT glosses.
The panel that shows which speakers (of the frequency data) appear in which documents shows that there isn't much data.
Refreshing does nothing unless you scroll around in the logs (not visible to the user yet) to see what glosses it couldn't find and create some more glosses.
(This doesn't work of course because I can only create glosses for annotation that haven't been found, not for gloss ids (of NGT glosses).
This example is meant to motivate the need for having extra buttons and the "appears in documents".
I intend to display what glosses it doesn't find in a feedback panel.
It's also possible to add  import more eaf files (via the Dataset Management) and then process those via "Refresh Documents". (That needs some more work.)


![corpus_overview_tstMH](https://user-images.githubusercontent.com/24272544/140090754-a236e782-0faf-497f-a137-21811880d39b.png)

**Comment by susanodd** [comment]
And here are two datasets for which there is no corpus. Kata Kolok has some (bogus) metadata and one eaf file.
And Norwegian Sign Language has no files yet. It only gets helpful information.


![corpus_overview_katakolok](https://user-images.githubusercontent.com/24272544/140091874-e48073b6-27cf-43a0-aba9-30fe382012bd.png)


![corpus_overview_NTS](https://user-images.githubusercontent.com/24272544/140091909-c3fbd137-f60d-4a0a-8db9-3d5511822490.png)

**Comment by susanodd** [comment]
To reflect that an eaf file may not yet be annotated, some additional columns have been added to the corpus overview. Namely, Number of Glosses, and a Comment column.


![corpus_ngt_overview_colms](https://user-images.githubusercontent.com/24272544/141127330-8da15cfd-68d7-41e1-a224-f906045e8b46.png)

There is also a column Updated to show whether the eaf file on disk is newer than the document that is currently (processed) in the corpus.
Additionally, if new eaf files are found that are not yet included in the corpus they are shown below the Documents table under New EAF Files as shown below.

![corpus_tstMH_overview_colms](https://user-images.githubusercontent.com/24272544/141127369-762ee9ae-6ffa-466b-a2ce-09546d6a72dc.png)

Some additional buttons will be added for processing new eaf files (as in the Documents table for refreshing documents).

During processing there is actually information available about what glosses that are in the eaf files are not found in the dataset.
This information would assist the user in knowing glosses are missing from their dataset.

The motivation for doing this is to reduce processing time. Here the user can see which files have been updated or have been uploaded and are not yet included.
The user can also spot which files are not yet annotated.

To process all 2281 CNGT eaf files takes lots of time. The goal of the setup shown above is to allow the user to selectively process files. However, the button for processing all files is still shown. This will probably need to become a background task request.

**Comment by susanodd** [comment]
I made the rows with Refresh buttons be Ajax calls to speed things up.
The last column is used for feedback. (In the screenshot, document CNGT0004 has been refreshed.)

![corpus-ngt-overview-ajax-buttons](https://user-images.githubusercontent.com/24272544/143578447-5d6cdd18-5f42-4199-9328-41964aa96fc0.png)

![corpus-ngt-overview-ajax-buttons-feedback](https://user-images.githubusercontent.com/24272544/143578469-6706ec84-4e71-47e2-98b3-5035e5687942.png)

There is actually a lot of data available that could potentially be shown to the user.

One thing I encountered is that if the "current document stored in the database" has glosses (GlossFrequency objects for the Gloss, Speaker, Document combination), and that Gloss is no longer found in Signbank, the GlossFrequency object should be deleted. I don't know how this happened, but in document CNGT0004, six of its glosses are no longer in Signbank NGT, but there are still GlossFrequency objects stored. The inconsistency was discovered because the "Number of Glosses" was not matching the number of updated glosses (i.e., Annotations found in the document) being shown in the log messages.

That is kind of also the intention of the Refresh buttons. Because it can happen that Glosses are either added or removed from the Dataset. So the Gloss might not exist anymore, or a Gloss might now be found.

**Comment by susanodd** [comment]
The latest revisions using Ajax calls have been pushed to branch corpus_functionality.

**Comment by susanodd** [comment]
@Woseseltops can you check whether the new server has the EAF files ported?
Related to this issue, the urls for the corpora are on the live server, but there are currently no EAF files for NGT.
However this cannot be true because there is frequency data for NGT. The files should be in writable/eafs/DATASET/
and the meta data files in writable/metadata_eafs/   (directory where the eafs are uploaded to, there should at least be one for NGT). Can you port them from the old server?

**Comment by susanodd** [comment]
Once upon a time, when the GlossFrequency model infrastructure was created, it was populated with the eaf files for NGT. So the current frequency data in signbank is from those eaf files that were once loaded.

**Comment by susanodd** [comment]
@Woseseltops can you check whether the EAF files have been ported to the new server?

writable/eafs/DATASET/
and the meta data files in 
writable/metadata_eafs/

**Comment by susanodd** [comment]
The cron job now works on the uploaded eaf files.
This supports the submaps too.

**Comment by susanodd** [comment]
Hmm. On the corpus overview page for tstMH clicking on the button to refresh a document seems to go to a wrong address as viewed in the Network debug panel of the browser.
